csharp/src/Google.Protobuf/Collections/RepeatedField.cs

```diff
@@ -320,10 +341,20 @@ public void Add(IEnumerable<T> values)
             {
                 throw new ArgumentNullException("values");
             }
-            // TODO: Check for ICollection and get the Count, to optimize?
-            foreach (T item in values)
+
+            ICollection collection = values as ICollection;
+            if (collection != null)
             {
-                Add(item);
+                this.EnsureSize(this.count + collection.Count);
+                collection.CopyTo(this.array, this.count);
```

Comment:
Using different implementations definitely complicates other aspects. I had something similar at one point, and while I can't remember the details, I was pleased to get back to a single implementation.

By "Let's ignore `int?` `float?` etc for now" do you mean "let's not optimize for those cases"? If so, I agree - but we can't "ignore" them in terms of "not caring whether the result is correct or not".

I'd be happy to make the new PR myself if you'd prefer - in particular, I'd branch it off another PR I have in flight for AddRange.